accepted/0000-registry-spec.md

```diff
@@ -0,0 +1,160 @@
+and optionally SemVer range or dist-tag.
+
+## Motivation
+
+Occasionally, users wish to use multiple npm registries.  For example, they
+may have some packages hosted on the public npm registry, and others within
+a private registry on their company's intranet, or provided by a company
+like GitHub, Jfrog, Sonatype, or others.
+
+Currently, it is possible to map a scope to a given registry, and all
+packages starting with that scope will be published to and installed from
+the defined registry:
+
+```ini
+; .npmrc file
+@company:registry = https://npm-registry.my-company.com
+```
+
+However, this does not address the following use cases:
+
+- Users publish a package to one registry as a "staging" area, and then
+  later wish to "upgrade" it to the public npm registry for broader
+  distribution.  Some code should fetch from the internal registry (for
+  example, for testing and validation prior to promotion), but this is
+  challenging unless the private registry also can serve _any_ public npm
+  package.
+
+- Users have a set of unscoped package dependencies, some of which come
+  from the public registry, and others which have patches applied to them
+  (either to the code, or to the packument to add warnings via the
+  `deprecated` field for example).  Again, this is only feasible right now
+  by making the internal registry capable of proxying the entire set of npm
+  packages.
+
+- Alias package specifiers cannot point to any registries other than the
+  primary `--registry` configuration.  It would be useful in many scenarios
+  to be able to alias a package to a copy found on a different registry.
+
+- Migrating packages from one registry to another is particularly
+  challenging, requiring downloading the tarball locally and then
+  re-uploading it.  It would be much simpler to script such migrations by
+  being able to do `npm publish <pkg from src registry> --registry=<dest
+  registry>`.
+
+- Using multiple different registries in the best of cases always requires
+  using a `.npmrc` file for configuration, which is a challenge in some
+  environments, and always adds a layer of indirection.  It would be more
+  straightforward at times to specify the registry right in the
+  package.json file or command line.
+
+## Detailed Explanation
+
+A new dependency specifier is added:
+
+```
+registry:<registry url>#<package name>[@<specifier>]
+```
+
+Where:
+
+- `<registry>` is a fully qualified URL to an npm registry, which may not
+  contain a `hash` portion,
+- `<package name>` is the (scoped or unscoped) name of the package to
+  resolve on the registry, and
+- `<specifier>` is an optional dist-tag, version, or range.
+
+If `<specifier>` is omitted, then it defaults to the `tag` config (or
+`defaultTag` internal optional), which defaults to `latest`.
+
+### Saving
+
+When a package is installed using a registry specifier, it *must* be saved
+using a registry specifier.
```

Comment:
If I run `npm install foo@registry:https://registry.foo.com/#foo@1.x`, and it is saved in such a way that future `npm install` invocations do not fetch from `https://registry.foo.com/`, then that is a clear violation of expectation and intent.

Reply:
Precisely; and if you run `npm install foo@registry:openjs#foo@1.x` then there's no such expectation, and users will instead assume they can swap the `openjs` registry url to another if they need - which is a reasonable feature, considering that it's already an established use case (cf the Taobao example).

Reply:
What this is saying is that we have to save it to the `package.json` so that if you run this:

```
npm install foo@registry:https://registry.foo.com/#foo
rm -rf node_modules
npm install
```

then the second one gets it from the `https://registry.foo.com` registry, and not the default configured registry.

If you agree, then I don't understand the objection.